driver-core/src/main/com/mongodb/ConnectionString.java

```diff
@@ -643,6 +643,9 @@ private MongoCredential createMongoCredentialWithMechanism(final AuthenticationM
         switch (mechanism) {
             case GSSAPI:
                 credential = MongoCredential.createGSSAPICredential(userName);
+                if (!authSource.equals("$external") && database != null) {
```

Comment:
I'm not clear why the value of `database` is relevant here.

Reply:
because the default value for authSource is "admin" not null, so if database is not null then the user explicitly set a value.

Reply:
Changing the URI in the "should throw an exception when an authMechanism (GSSAPI) has an authSource other than $external" test to:

    mongodb://user%40DOMAIN.COM@localhost/?authMechanism=GSSAPI&authSource=bar

makes the test fail (i.e. it no longer throws an exception).  Shouldn't it still succeed?

Reply:
Yes it should have.  I have changed how the authSource is defaulted.

Reply:
OK, this is interesting.  Looking at the history of connection-string.json I see this test

    {
      "description": "should use $external as the source when a database is specified",
      "uri": "mongodb://user%40DOMAIN.COM:password@localhost/foo?authMechanism=GSSAPI",
      "hosts": [],
      "valid": true,
      "warning": false,
      "auth": {
        "db": "$external"
      },
      "options": null
    }

which doesn't appear in this pull request (the new test explicitly specifies authSource=$external for GSSAPI).  Prior to the ConnectionString logic changes in this PR, the Java driver would have passed this test, but now it would fail.  Looking at the auth spec, the new logic and associated test in this PR look strictly correct, but it's a backwards breaking change for applications.  Was that why you made this change, to be consistent with the spec?

Reply:
I saw that test as invalid on two counts - one the wrong auth db and second it includes an unused password.

Reply:
Yeah, it's not a good test because of the password... but this new requirement of an explicit authSource=$external for GSSAPI etc. in all circumstances is going to break existing applications.  We've never documented it, and never enforced it, so I'm pretty sure no one is doing it.

Reply:
Well its correct as per both the connection string spec **and** the auth spec. Its not an explict requirement for authSource=$external, its only explicit if the auth database is also used for an alternative purpose. Both urls provide the same authSource:

```
mongodb://user%40DOMAIN.COM@localhost/foo?authMechanism=GSSAPI
mongodb://user%40DOMAIN.COM@localhost/authMechanism=GSSAPI&authSource=foo
```

I'm not sure the initial tests weren't ever vetted, certainly doesn't look like many drivers have implemented them - but the drivers ticket is to implement the tests, which this PR does.  

I'm happy to remove the checks out, but the driver will be out of Spec. Alternatively, we can ditch this PR and mark the ticket as won't fix?